File: examples/pyomo/pmedian.py
# TODO

import random
random.seed(1000)

from pyomo.environ import *
from poek.util import pyomo_to_poek


#
# Create a p-median optimization problem.  
# * Variables are initialized to zero to suppress warnings in the pyomo_to_poek function.
# * Parameters are mutable to illustrate their capture in pyomo_to_poek.
#
def create_model():

    model = AbstractModel()

    model.N = Param(within=PositiveIntegers)

    model.Locations = RangeSet(1,model.N)

    model.P = Param(within=RangeSet(1,model.N))

    model.M = Param(within=PositiveIntegers)

    model.Customers = RangeSet(1,model.M)

    model.d = Param(model.Locations, model.Customers, initialize=lambda n, m, model : random.uniform(1.0,2.0), within=Reals, mutable=True)

    model.x = Var(model.Locations, model.Customers, bounds=(0.0,1.0), initialize=0)

    model.y = Var(model.Locations, within=Binary, initialize=0)

    def rule(model):
        return sum( model.d[n,m]*model.x[n,m] for n in model.Locations for m in model.Customers )
    model.obj = Objective(rule=rule)

    def rule(model, m):
        return (sum( model.x[n,m] for n in model.Locations ), 1.0)
    model.single_x = Constraint(model.Customers, rule=rule)

    def rule(model, n,m):
        return (None, model.x[n,m] - model.y[n], 0.0)
    model.bound_y = Constraint(model.Locations, model.Customers, rule=rule)

    def rule(model):
        return (sum( model.y[n] for n in model.Locations ) - model.P, 0.0)
    model.num_facilities = Constraint(rule=rule)

    return model


#
# Create the Pyomo model
#
model = create_model()
instance = model.create_instance(filename='pmedian.dat')
#
# Translate to POEK
#
pm = pyomo_to_poek(instance)
#
# Display the model
#
pm.poek_model.print_equations()
#
# Mappings for variables and parameters
#
print("Variables: id(vardata) -> poek-variable")
for key, vardata in pm.pyo2pk_var.items():
    print(key, vardata, vardata.name)
print("Parameters: id(paramdata) -> poek-parameter")
for key, paramdata in pm.pyo2pk_param.items():
    print(key, paramdata, paramdata.name)

File: test/aml_comparisons/coek/models/misc/pmedian.cpp
#include <coek/coek.hpp>
#include <map>
#include <vector>

void pmedian_scalar(coek::Model& model, size_t N, size_t P)
{
    size_t M = N;  // Customers

    std::vector<std::vector<double>> d;
    for (size_t n = 0; n < N; n++) {
        d.push_back(std::vector<double>(M));
        for (size_t m = 0; m < M; m++)
            d[n][m] = 1.0 + 1.0 / (n + m + 1);
    }

    std::vector<std::vector<coek::Variable>> x;
    for (size_t n = 0; n < N; n++) {
        x.push_back(std::vector<coek::Variable>(M));
        for (size_t m = 0; m < M; m++)
            model.add(x[n][m].bounds(0, 1).value(0));
    }

    std::vector<coek::Variable> y(N);
    for (size_t n = 0; n < N; n++)
        model.add(y[n].bounds(0, 1).value(0));

    // obj
    coek::Expression obj;
    for (size_t n = 0; n < N; n++)
        for (size_t m = 0; m < M; m++)
            obj += d[n][m] * x[n][m];
    model.add_objective(obj);

    // single_x
    for (size_t m = 0; m < M; m++) {
        coek::Expression c;
        for (size_t n = 0; n < N; n++)
            c += x[n][m];
        model.add(c == 1);
    }

    // bound_y
    for (size_t n = 0; n < N; n++)
        for (size_t m = 0; m < M; m++)
            model.add(x[n][m] - y[n] <= 0);

    // num_facilities
    coek::Expression num_facilities;
    for (size_t n = 0; n < N; n++)
        num_facilities += y[n];
    model.add(num_facilities == (int)P);
}
